sdk/dotnet/Pulumi.Automation/Events/EventLogWatcher.cs

```diff
@@ -0,0 +1,123 @@
+        private readonly CancellationTokenSource _internalCancellationTokenSource = new CancellationTokenSource();
+        private readonly CancellationToken _externalCancellationToken;
+        private readonly Action<EngineEvent> _onEvent;
+        private readonly Task _pollingTask;
+
+        private const int _pollingIntervalMilliseconds = 100;
+
+        // We keep track of the length we have already read from the file
+        private long _previousLength = 0;
+
+        public string LogFile { get; }
+
+        internal EventLogWatcher(
+            string logFile
+            , Action<EngineEvent> onEvent
+            , CancellationToken externalCancellationToken)
+        {
+            LogFile = logFile;
+            _onEvent = onEvent;
+            _externalCancellationToken = externalCancellationToken;
+            _pollingTask = Task.Run(PollForEvents);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            _internalCancellationTokenSource.Cancel();
+            await _pollingTask.ConfigureAwait(false);
+            _internalCancellationTokenSource.Dispose();
+        }
+
+        private async Task PollForEvents()
+        {
+            var internalToken = _internalCancellationTokenSource.Token;
+            var externalToken = _externalCancellationToken;
+
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(internalToken, externalToken);
+
+            var linkedToken = linkedSource.Token;
+
+            while (!linkedToken.IsCancellationRequested)
+            {
+                try
+                {
+                    // NOTE: Waiting before reading so that if ReadEvents throws we will
+                    // wait before attempting it again
+                    await Task.Delay(_pollingIntervalMilliseconds, linkedToken).ConfigureAwait(false);
+                    ReadEvents();
+                }
+                catch (TaskCanceledException)
+                {
+                    break;
+                }
+                catch
+                {
+                    // We do our best effort to keep reading until the polling is cancelled
+                }
+            }
+
+            // Try reading events one final time
+            try
+            {
+                ReadEvents();
+            }
+            catch
+            {
+                // ignored
+            }
+        }
+
+        private void ReadEvents()
+        {
+            if (!File.Exists(LogFile))
+            {
+                return;
+            }
+
+            using var fs = new FileStream(LogFile, FileMode.Open, FileAccess.Read);
+            var newLength = fs.Length;
+
+            if (newLength == _previousLength)
+            {
+                return;
+            }
+
+            fs.Seek(_previousLength, SeekOrigin.Begin);
+
+            using var reader = new StreamReader(fs);
+
+            var lines = reader
```

Comment:
This bit me on a badly behaved test, not sure if it's super relevant but worth considering. A fast running chatty command will be generating lots of data here, which will require O(N) ram as we load it all up, moreover, we do not process the initial events until we finish ReadToEnd. 

An alternative would be to loop over the StreamReader with ReadLine or ReadLineAsync.